zaza/openstack/charm_tests/cinder_lvm/tests.py

```diff
@@ -0,0 +1,136 @@
+        cls.cinder_client = openstack_utils.get_cinder_session_client(
+            cls.keystone_session)
+
+    @classmethod
+    def tearDown(cls):
+        """Remove test resources."""
+        volumes = cls.cinder_client.volumes
+        for volume in volumes.list():
+            if volume.name.startswith('zaza'):
+                try:
+                    volume.detach()
+                    volumes.delete(volume)
+                except Exception:
+                    pass
+
+    def test_cinder_config(self):
+        """Test that configuration options match our expectations."""
+        logging.info('cinder-lvm')
+        expected_contents = {
+            'LVM-zaza-lvm': {
+                'volume_clear': ['zero'],
+                'volumes_dir': ['/var/lib/cinder/volumes'],
+                'volume_name_template': ['volume-%s'],
+                'volume_clear_size': ['0'],
+                'volume_driver': ['cinder.volume.drivers.lvm.LVMVolumeDriver'],
+            }}
+
+        zaza.model.run_on_leader(
+            'cinder',
+            'sudo cp /etc/cinder/cinder.conf /tmp/',
+            model_name=self.model_name)
+        zaza.model.block_until_oslo_config_entries_match(
+            'cinder',
+            '/tmp/cinder.conf',
+            expected_contents,
+            model_name=self.model_name,
+            timeout=10)
+
+    def _create_volume(self):
+        """Create a volume via the LVM backend."""
+        test_vol_name = "zaza{}".format(uuid.uuid1().fields[0])
+        vol_new = self.cinder_client.volumes.create(
+            name=test_vol_name,
+            size='1')
+        openstack_utils.resource_reaches_status(
+            self.cinder_client.volumes,
+            vol_new.id,
+            wait_iteration_max_time=12000,
+            stop_after_attempt=5,
+            expected_status='available',
+            msg='Volume status wait')
+        return self.cinder_client.volumes.find(name=test_vol_name)
+
+    def test_create_volume(self):
+        """Test creating a volume with basic configuration."""
+        test_vol = self._create_volume()
+        self.assertTrue(test_vol)
+
+        host = getattr(test_vol, 'os-vol-host-attr:host').split('#')[0]
+        self.assertTrue(host.startswith('cinder@LVM'))
+
+    @with_conf('cinder-lvm', {'overwrite': 'true', 'block-device': '/dev/vdc'})
+    def test_volume_overwrite(self):
+        """Test creating a volume by overwriting one on the /dev/vdc device."""
+        self._create_volume()
+
+    @with_conf('cinder-lvm', {'block-device': 'none'})
+    def test_device_none(self):
+        """Test creating a volume in a dummy device (set as 'none')."""
+        self._create_volume()
+
+    @with_conf('cinder-lvm', {'remove-missing': 'true'})
+    def test_remove_missing_volume(self):
+        """Test creating a volume after remove missing ones in a group."""
+        self._create_volume()
+
+    @with_conf('cinder-lvm', {'remove-missing-force': 'true'})
```

Comment:
IIUC this under the hood should end up calling `vgreduce --removemissing ...`, I think you may need to setup a scenario where there are VGs referencing missing PVs so we are certain that under those circumstances it's possible to create volumes.

Reply:
@freyes as discussed a couple days ago, cinder makes it hard to actually test this in a clean manner, since only a single volume name can be configured as opposed to a comma-separated list. We could, of course, manually create them in the tests, but it will end up polluting the test environment, since all the LVM stuff are global settings and can't be isolated.

Reply:
I think this is a good test to have. Yes, it will manipulate the test environment - but that can be done as the last test to run so it doesn't inadvertently pollute other tests that are exercised. However, I think I'm comfortable with this being a subsequent improvement to the cinder-lvm tests as long as @lmlg is willing to commit to do those in a follow-on patch.

Reply:
I can absolutely commit them in a follow up PR.